File: frontend/lib/models/member.dart
import 'dart:convert';

import 'package:frontend/models/contact.dart';
import 'package:frontend/models/team.dart';

class Member {
  final String id;
  final String name;
  final String? image;
  final String istId;
  final String sinfoId;
  final String? contact;

  Member({
    required this.id,
    required this.name,
    this.image,
    required this.istId,
    required this.sinfoId,
    this.contact,
  });

  factory Member.fromJson(Map<String, dynamic> json) {
    return Member(
      id: json['id'],
      name: json['name'],
      image: json['img'],
      istId: json['istid'],
      sinfoId: json['sinfoid'],
      contact: json['contact'],
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'img': image,
        'istid': istId,
        'sinfoid': sinfoId,
        'contact': contact,
      };

  @override
  String toString() {
    return json.encode(this.toJson());
  }

  bool operator ==(o) => o is Member && id == o.id;
  int get hashCode => id.hashCode;
}

class MemberPublic {
  final String? name;
  final String? image;
  final ContactSocials? socials;

  MemberPublic({
    this.name,
    this.image,
    this.socials,
  });

  factory MemberPublic.fromJson(Map<String, dynamic> json) {
    return MemberPublic(
      name: json['name'],
      image: json['img'],
      socials: ContactSocials.fromJson(json['socials']),
    );
  }

  Map<String, dynamic> toJson() => {
        'name': name,
        'img': image,
        'socials': socials?.toJson(),
      };
}

class MemberParticipation {
  final int? event;
  final String? role;
  final String? team;

  MemberParticipation({
    this.event,
    this.role,
    this.team,
  });

  factory MemberParticipation.fromJson(Map<String, dynamic> json) {
    return MemberParticipation(
      event: json['event'],
      role: json['role'],
      team: json['team'],
    );
  }

  Map<String, dynamic> toJson() => {
        'event': event,
        'role': role,
        'team': team,
      };
}

File: backend/src/models/member.go
package models

import (
	"go.mongodb.org/mongo-driver/bson/primitive"
)

// Member represents a member of the SINFO organization
type Member struct {
	// Member's ID (_id of mongodb).
	ID primitive.ObjectID `json:"id" bson:"_id"`

	Name string `json:"name" bson:"name"`

	// Photo of the member. This is a URL pointing to the image, that is
	// being stored somewhere else.
	Image string `json:"img" bson:"img"`

	// Instituto Superior Técnico's Identification number of the member.
	// Can be left blank if non-applicable.
	ISTID string `json:"istid" bson:"istid"`

	// SINFOID is the member's identifier ID under the SINFO organization
	// Example: if the member's SINFO email is john.doe@sinfo.org, his sinfo ID would be john.doe
	SINFOID string `json:"sinfoid" bson:"sinfoid"`

	// Contact is an _id of Contact (see models.Contact).
	Contact primitive.ObjectID `json:"contact" bson:"contact"`
}

// MemberPublic is the public information about a member
type MemberPublic struct {
	Name string `json:"name" bson:"name"`

	// Photo of the member. This is a URL pointing to the image, that is
	// being stored somewhere else.
	Image string `json:"img" bson:"img"`

	Socials ContactSocials `json:"socials"`
}

type AuthorizationCredentials struct {
	// Member's ID
	ID primitive.ObjectID

	// SINFO's email ID (example: if your email is john.doe@sinfo.org, your emailID would be john.doe)
	SINFOID string

	// Role on SINFO
	Role TeamRole
}

type MemberEventTeam struct {
	//Number of event where Member was in <Team>
	Event int

	//Team of Member on SINFO <Event>
	Team string

	// Role on SINFO <Event>
	Role TeamRole
}
